### ISSUE TITLE
Details/summary polyfill (expandable/collapsible content):

### ISSUE THREAD

**Issue by Ferkner** [wet-boew/wet-boew#8633]
Somewhat related to https://github.com/wet-boew/wet-boew/issues/5141, can that method be used (Javascript onlick event) be used if the content is being linked to from another page?

Page A has a bunch of links that go to different summary elements on Page B. When someone clicks on a link on Page A, I would like it to take them to the right spot on Page B and have that summary element automatically opened.

**Comment by thekodester** [comment]
You need to create some customer JavaScript, I'm doing this on a blog I run.

```
(function() {
  const stopAnimation = function () {
    if ( $(  'html, body'  ).is( ':animated' ) ) {
      $('html, body').stop();
    }
  };
  window.addEventListener('mousewheel', stopAnimation, false);
  window.addEventListener('DOMMouseScroll', stopAnimation, false);

  if( window.location.hash ) {
    const hash = window.location.hash.replace( /^\#/, '' ),
      element = document.getElementById( hash );

    if ( isElement( element ) ) scrollTop( element );
  }

  window.addEventListener( 'hashchange', function() {
    if ( window.location.pathname.includes( 'legal' ) ) {
      const hash = location.hash.replace( /^\#/, '' )

      let element, details;
      if ( hash ) element = document.getElementById( hash );
      if ( isElement( element ) && element.tagName.toLowerCase() === 'details' ) details = element;
      if ( isElement( element ) && element.parentElement.tagName.toLowerCase() === 'details' ) details = element.parentElement;

      if ( isElement( details ) ) {
        details.open = true;
        scrollTop( element );
      }
    }
  }, false);

  document.querySelectorAll( 'a[data-slider="true"]' ).forEach(anchor => {
    anchor.addEventListener( 'click', function ( e ) {
      const hash = this.getAttribute( 'href' ).replace( /^\/\#/, '' ),
        element = document.getElementById( hash );

      if ( isElement( element ) ) {
        e.preventDefault();
        scrollTop( element );
      }
    });
  });

  function scrollTop( element ) {
    $( 'html, body' ).animate({
      scrollTop: ( $( element ).offset().top - 65 )
    }, 1000, "easeInOutExpo");
  }

  function isElement( element ) {
    return ( element != null && typeof( element ) != 'undefined' );
  }
})();
```
You can see the code in action here: https://recipeapproved.ca/legal/.

Also you can see the page open and scroll to a details/summary here: https://recipeapproved.ca/legal/#rights

**Comment by Ferkner** [comment]
All your links are on the same page as the content. How would this work coming in from a different page to a specific summary element?

**Comment by thekodester** [comment]
If you click on the link with a hashtag (in the comment above) it goes to that part of the summary from here on GitHub.

**Comment by Ferkner** [comment]
But the Privacy Policy is already open. I want to be able to link to a page like http://www.agr.gc.ca/eng/?id=1535571735990 and have it auto open only the summary linked to.

**Comment by thekodester** [comment]
The code does that `details.open = true;`;

**Comment by thekodester** [comment]
You'll need to modify it to your needs, but that not my job sorry.

**Comment by Ferkner** [comment]
I know it's not. thanks for this. I'll see if I can get it working :)

**Comment by thekodester** [comment]
The code is a start pointing. I wouldn't use a GET request, but hashtags instead. If you want to use a GET request I would look at https://stackoverflow.com/questions/979975/how-to-get-the-value-from-the-get-parameters.

**Comment by thekodester** [comment]
I'll try also adding some comments to code make it easier to read.

**Comment by thekodester** [comment]
Comments where added.

**Comment by Ferkner** [comment]
Thank you!

**Comment by thekodester** [comment]
I modified the code a bit and closed the summary on page to show the code working another way. If you go here it opens the summary and scrolls to it content https://recipeapproved.ca/legal/#dataretention.

**Comment by juleskuehn** [comment]
To clarify:

- You want to open a specific summary when linked to.
- Your link would end in a hash fragment, ex. [#wb-auto-12](http://www.agr.gc.ca/eng/about-us/planning-and-reporting/service-standards-at-agriculture-and-agri-food-canada/quarterly-results-fiscal-year-2018-2019/?id=1535571735990#wb-auto-12)
- The above link, by default browser behaviour, will scroll to the element with the ID of the hash fragment (wb-auto-12).
- You want this detail summary set to "open".

If this is correct, you simply need to include the code below on the page linked to. Note that if you have an existing handler for window.onload you should add the function body into that handler...

```
window.onload = function () {
  if (window.location.hash) {
    document.getElementById(window.location.hash.substring(1)).setAttribute('open', '');
  }
}
```

This could be implemented into the WET detail/summary plugin.

**Comment by thekodester** [comment]
@juleskuehn I disagree with adding to the detail/summary plugin as it could lead to accessibility issues. For one there is no context to the user of what was opened as it doesn't scroll to the element. Also your code has no check to make sure the element exist and that it is a detail/summary.

The code I posted already does this:
-checks if detail/summary
-checks if within details/summary
-scrolls to element that was opened

**Comment by juleskuehn** [comment]
Hm, I see that it indeed does not scroll to the element in my linked example!

Apart from that, what are the accessibility issues?

I am trying to find a simpler solution because it would be valuable to have this behaviour across the board (incorporated into WET), and don't want to add anything not strictly required (animations, mousewheel handlers).

Regarding the parent elements - your code does check if the element is a direct descendant of a `<details>` element, but there is no guarantee this would be the case. We could use a loop to ensure that all parent `<details>` elements are opened.
```
let scrollToHash = function () {
  if (window.location.hash) {
    let el = document.getElementById(window.location.hash.substring(1));
    if (el === null) return;
    // Open any parent <details> elements
    for (let elem= el; elem !== document; elem = elem.parentNode) {
      if (elem.tagName.toLowerCase() === 'details') {
        elem.setAttribute('open', '');
      }
    }
    el.scrollIntoView();
  }
}

window.onload = scrollToHash;
```

Thoughts?

**Comment by Ferkner** [comment]
The following did not work for me: 
`window.onload = function () {
  if (window.location.hash) {
    document.getElementById(window.location.hash.substring(1)).setAttribute('open', '');
  }
}`

Because the 'open' didn't work, I tried changing that line to

`document.getElementById(window.location.hash.substring(1)).click();`

hoping that simulating a mouse clock would work. But it did not. Now, I know almost nothing about JavaScript, so I am quite out of my element. Each details/summary has a unique ID, and they are linked to from a parent page so when you click on a link on the parent page you go to the right spot on the child page with all the details/summary. If I could get the mouseclick thing to work, I would be set. Currently, this is the code I have at the bottom of the body:

`<script>
window.onload = function () {
  if (window.location.hash) {
    document.getElementById(window.location.hash.substring(1)).setAttribute('click', '');
  }
}
</script>`

**Comment by juleskuehn** [comment]
Have you tried the second (longer) code snippet I posted last? If that doesn't work, can you link to the page you are working on?

**Comment by Ferkner** [comment]
I ended up getting it working. The problem was that my ids were in the <summary> tag. Once I moved them to the <h2> that appears within the <summary> it worked. It's doing exactly what I want, so a thousand thank yous!

Those those who come here in the future, here is some code to get it working:

`<section id="details-elements">
  <div class="btn-group mrgn-bttm-lg">
    <button type="button" class="btn btn-default wb-toggle mrgn-rght-sm" data-toggle='{"selector": "details", "parent": "#details-elements", "print": "on", "type": "on"}'>Display All</button>
    <button type="button" class="btn btn-default wb-toggle mrgn-lft-sm" data-toggle='{"selector": "details", "parent": "#details-elements", "type": "off"}'>Collapse All</button>
  </div>
  <details>
    <summary>
      <h2 class="h4" id="ar">Arabic <i lang="ar">العربية</i></h2>
    </summary>
  <p>content</p>
  </details>
  <details>
    <summary>
      <h2 class="h4" id="zhs">Chinese (Simplified) <i lang="zh">中文</i></h2>
    </summary>
  <p>content</p>
  </details>
  <details>
    <summary>
      <h2 class="h4" id="zht">Chinese (Traditional) <i lang="zh">中文</i></h2>
    </summary>
  <p>content</p>
  </details>
  <details>
    <summary>
      <h2 class="h4" id="hr">Croatian <i lang="hr">Hrvatski</i></h2>
    </summary>
  <p>content</p>
  </details>
  <details>
    <summary>
      <h2 class="h4" id="nl">Dutch <i lang="nl">Nederlands</i></h2>
    </summary>
  <p>content</p>
  </details>
  <details>
    <summary>
      <h2 class="h4" id="ko">Korean <i lang="ko">한국어</i></h2>
    </summary>
  <p>content</p>
  </details>
  <details>
    <summary>
      <h2 class="h4" id="ms">Malay <i lang="ms">Melayu</i></h2>
    </summary>
  <p>content</p>
  </details>
</section>

<script>
window.onload = function () {
  if (window.location.hash) {
    document.getElementById(window.location.hash.substring(1)).click();
  }
}
</script>`

**Comment by Ferkner** [comment]
I ended up getting it working. The problem was that my ids were in the summary tag. Once I moved them to the h2 that appears within the summary it worked. It's doing exactly what I want, so a thousand thank yous!

I wanted to include a bit of code for future seekers who might be looking for something like this, but it doesn't leave my HTML code as code within the code  markers.

Basically, your first tag after summary, whether it's a `<p>` tag or an `<h2>` tag, needs to have the id that will act as the anchor. Then at the bottom of your content you put the following code:

`<script>
window.onload = function () {
  if (window.location.hash) {
    document.getElementById(window.location.hash.substring(1)).click();
  }
}
</script>`

That should get it working for you like it has for me.
